crypto/evp/m_sha3.c

```diff
@@ -404,3 +410,27 @@ EVP_MD_SHA3(512)
 
 EVP_MD_SHAKE(128)
 EVP_MD_SHAKE(256)
+
+
+# define EVP_MD_KMAC(bitlen)                    \
+const EVP_MD *EVP_kmac##bitlen(void)            \
+{                                               \
+    static const EVP_MD kmac_##bitlen##_md = {  \
+        NID_kmac##bitlen,                       \
```

Comment:
Does the NID need to be specified?  I'm presuming these are internal only functions since some of the KMAC processing is performed outside of this.

Does the EVP_MD code even work with a 0 or -1 NID specified???

This could avoid defining the (otherwise unused & unhelpful) kmac128 and kmac256 in objects.txt.

Reply:
NIDs can certainly be avoided, just don't call `EVP_add_digest` with these structures, 'cause the object database depends on proper NIDs.  This *does* mean, however, that `EVP_get_bydigestname` can't be used in `kmac_ctrl_str`, and you'll have to do with hard-coded MD algo strings.  I would be fine with that, considering it's limited to KMAC128 or KMAC256 anyway...

Reply:
Pauli doing this causes a few issues in the evp_test.
i.e- I can no longer use the 'Algorithm' test field as this does the  EVP_get_bydigestname(). (Unless I start doing the KMAC checks here).
Taking this out and using a ctrl = digest:KMAC128 instead means that the key gets set before the digest which is problematic. Rearranging the mac_test code so that the ctrls get set up before the SET_KEY is done works.

BUT...
then one of the Siphash tests functions fails...
There seems to be something a bit fishy in the siphash ctrls...
Changing the set_key / set_size order causes an issue when digest_size = 8)
i.e: SIPHASH_MIN_DIGEST_SIZE (That looks like a side effect to me).

static int siphash_ctrl(EVP_MAC_IMPL *sctx, int cmd, va_list args)
{
    switch (cmd) {
    case EVP_MAC_CTRL_SET_SIZE:
               return SipHash_set_hash_size(&sctx->ctx, size);
        }
        break;
    case EVP_MAC_CTRL_SET_KEY:
                return SipHash_Init(&sctx->ctx, key, 0, 0);
        }
        break;
}

/* hash_size = crounds = drounds = 0 means SipHash24 with 16-byte output */
int SipHash_Init(SIPHASH *ctx, const unsigned char *k, int crounds, int drounds)
{
    /* If the hash size wasn't set, i.e. is zero */
    ctx->hash_size = siphash_adjust_hash_size(ctx->hash_size);
    …..
    
    if (ctx->hash_size == SIPHASH_MAX_DIGEST_SIZE)
        ctx->v1 ^= 0xee;

    return 1;
}

static size_t siphash_adjust_hash_size(size_t hash_size)
{
    if (hash_size == 0)
        hash_size = SIPHASH_MAX_DIGEST_SIZE;
    return hash_size;
}

int SipHash_set_hash_size(SIPHASH *ctx, size_t hash_size)
{
    hash_size = siphash_adjust_hash_size(hash_size);
    if (hash_size != SIPHASH_MIN_DIGEST_SIZE
        && hash_size != SIPHASH_MAX_DIGEST_SIZE)
        return 0;

    ctx->hash_size = hash_size;
    return 1;
}

Reply:
Okay, too hard :)

How about the path proposed by @levitte instead.

Reply:
It's quite possible that the siphash method could use a bit of caching. I'm trying to avoid that, but it's not optimal either if it becomes too sensitive to ctrl order... maybe. I will have to look at that.

Reply:
Hi Richard,
To test this go into mac_test_run_mac() and change the order to

    if (algo_ctrl != 0) {
        rv = EVP_MAC_ctrl(ctx, algo_ctrl, algo);
        ....
    }

    for (i = 0; i < sk_OPENSSL_STRING_num(expected->controls); i++) {
       ....
    }

    rv = EVP_MAC_ctrl(ctx, EVP_MAC_CTRL_SET_KEY, expected->key, expected->key_len);

    if (expected->iv != NULL) {
        rv = EVP_MAC_ctrl(ctx, EVP_MAC_CTRL_SET_IV, expected->iv, expected->iv_len);
    }
    if (!EVP_MAC_init(ctx)) {

Reply:
So in the working case, where the key is set first
In the init() the hash_size is unset initially so it sets it to MAX and then changes ctx->v1
   │126         if (ctx->hash_size == SIPHASH_MAX_DIGEST_SIZE)                                                
   │127             ctx->v1 ^= 0xee;  
Then it sets the hash_size via a ctrl SET_SIZE which sets it to a size of 8

In the non working case SET_SIZE gets called which sets ctx->hash_size to 8. Then in the setKey the  Init wont modify ctx->v1...

That if on line 126 is very strange.. Maybe a flag could be set?

Reply:
I know diddly squat about siphash, I've only observed what it does, and what I meant to say is that `siphash_meth.c` may need some adjustment to be more flexible.  Remind me next week...

Reply:
According the test vectors in https://131002.net/siphash/siphash24.c the value we have in our test vectors is incorrect.

it should be 
# SIPHASH - default values: 2,4 rounds, explicit 8-byte mac

MAC = SipHash
Ctrl = digestsize:8
Key = 000102030405060708090A0B0C0D0E0F
Input = 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E
Output = 724506eb4c328a95

Reply:
I have updated the code to use EVP_MAC_KMAC128 & EVP_MAC_KMAC256 for the selection process.. The internal EVP_kmacX have been renamed to evp_keccak_kmacX